server/src/test/java/org/apache/calcite/avatica/server/CustomAuthHttpServerTest.java

```diff
@@ -0,0 +1,300 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.avatica.server;
+
+import org.apache.calcite.avatica.ConnectionSpec;
+import org.apache.calcite.avatica.jdbc.JdbcMeta;
+import org.apache.calcite.avatica.remote.AuthenticationType;
+import org.apache.calcite.avatica.remote.Driver;
+import org.apache.calcite.avatica.remote.LocalService;
+
+import org.eclipse.jetty.security.ConstraintSecurityHandler;
+import org.eclipse.jetty.security.UserAuthentication;
+import org.eclipse.jetty.server.Authentication;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.UserIdentity;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.sql.SQLException;
+import javax.servlet.http.HttpServletRequest;
+
+import static org.hamcrest.core.StringContains.containsString;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+import java.util.Arrays;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+/**
+ * Test class for providing CustomAvaticaServerConfiguration to the HTTP Server
+ */
+public class CustomAuthHttpServerTest extends HttpAuthBase {
+  private static final ConnectionSpec CONNECTION_SPEC = ConnectionSpec.HSQLDB;
+  private static HttpServer server;
+  private static String url;
+
+  private static int methodCallCounter1 = 0;
+  private static int methodCallCounter2 = 0;
+  private static int methodCallCounter3 = 0;
+
+  @Before
+  public void before() throws SQLException {
+    methodCallCounter1 = 0;
+    methodCallCounter2 = 0;
+    methodCallCounter3 = 0;
+  }
+
+  @Test
+  public void testCustomImpersonationConfig() throws Exception {
+    final JdbcMeta jdbcMeta = new JdbcMeta(CONNECTION_SPEC.url,
+            CONNECTION_SPEC.username, CONNECTION_SPEC.password);
+    LocalService service = new LocalService(jdbcMeta);
+
+    server = new HttpServer.Builder()
+            .withCustomAuthentication(new CustomImpersonationConfig())
+            .withHandler(service, Driver.Serialization.PROTOBUF)
+            .withPort(0)
+            .build();
+    server.start();
```

Comment:
You need to ensure any server you start is stopped after the test method completes

Reply:
The `@After` annotated method on Line 292, ensure that server is stopped after every test case.
I could not use the `@Before` method to start the server, since every test required a different type of server. I have tried to refactor the code so that we can reuse the methods as required. Let me know if it requires further refactoring.

Reply:
+1, Will make it consistent.